### ISSUE TITLE
Add ES Module (ESM) support to the factory

### ISSUE THREAD

**Issue by codingjoe** [bigskysoftware/htmx#2270]
Hi there 👋,

Currently, the factory supports AMD, CJS and browser environments.
However, modern websites and browsers have proper module (ESM) support.

Let's say you want to do something like this:

```html
<!DOCTYPE html>
<html lang="en">
<head>
<title>HTMX via ESM</title>
</head>
<body>
<script type="module">
import htmx from 'https://unpkg.com/htmx.org'
htmx.logAll()
</script>
</body>
</html>
```

Your browser will tell you that `htmx.org` doesn't export a default. Similarly, something like this also does not work:

```javascript
import { logAll } from 'https://unpkg.com/htmx.org'
logAll()
```

Both are completely valid use cases and have, as of writing this, over [97% browser support](https://caniuse.com/es6-module).

I think adoption would be rather simple. Everything already works in NodeJS or rather CommonJS (CJS). All that's needed would be to extend the factory to export ESM modules in a browser.

If you want to take this one step further, we could also consider to properly defining entry points via NodeJS' [exports](https://nodejs.org/api/packages.html#exports). This would be particularly helpful if you wish to load extensions.

I would be happy to contribute a small patch.

Cheers!
Joe

**Comment by Telroshan** [comment]
Hey, it seems (I know nothing about all this so you tell me 😆) we have esm support in [v2 alpha](https://github.com/bigskysoftware/htmx/blob/v2.0v2.0/dist/htmx.esm.js), if you're willing to take a look and test if things work as expected?

**Comment by codingjoe** [comment]
> Hey, it seems (I know nothing about all this so you tell me 😆) we have esm support in [v2 alpha](https://github.com/bigskysoftware/htmx/blob/v2.0v2.0/dist/htmx.esm.js), if you're willing to take a look and test if things work as expected?

Nope, no luck. 2.0 doesn't export anything. And the extensions are not declared in the package.json file, either.

I happen to know a thing or two about ESM, I am happy to help. Just say the word and I will gladly submit a patch.

**Comment by 1cg** [comment]
in v2.0 we have a .esm.js file that does export htmx:

https://github.com/bigskysoftware/htmx/blob/f004978bbbf68befb49bfef7f17a2ec25ce35ae0/dist/htmx.esm.js#L3917

**Comment by codingjoe** [comment]
> in v2.0 we have a .esm.js file that does export htmx:
> 
> https://github.com/bigskysoftware/htmx/blob/f004978bbbf68befb49bfef7f17a2ec25ce35ae0/dist/htmx.esm.js#L3917

Yes, I saw that, but that's not how you export an ESM module ;)

**Comment by alexpetros** [comment]
> > in v2.0 we have a .esm.js file that does export htmx:
> > https://github.com/bigskysoftware/htmx/blob/f004978bbbf68befb49bfef7f17a2ec25ce35ae0/dist/htmx.esm.js#L3917
> 
> Yes, I saw that, but that's not how you export an ESM module ;)

Would you like to elaborate on this...? I too have been laboring under the delusion that one exported [ECMAScript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) with the `export` command.

**Comment by 1cg** [comment]
is it an anonymous function?  I thought it was a plain javascript object returned by the IIFE...

**Comment by codingjoe** [comment]
> is it an anonymous function? I thought it was a plain javascript object returned by the IIFE...

https://github.com/bigskysoftware/htmx/blob/f004978bbbf68befb49bfef7f17a2ec25ce35ae0/src/htmx.js#L1

It's exporting the outer wrapper, not the object. Both would not be leading to the desired result. The latter would allow at least something like `import * as htmx from 'htmx.org'`. But that's maybe not the way to go here. A default export might be the most elegant and intuitive solution for most.

**Comment by alexpetros** [comment]
> Please excuse my previous response. I didn't explain it well. Let me try again:

Appreciated! And don't worry about it.

> The latter would allow at least something like import * as htmx from 'htmx.org'.

Correct me if I'm wrong, but the current `export { htmx }` statement we have in the the `esm` wrapper supports this, right?

**Comment by codingjoe** [comment]
> > The latter would allow at least something like import * as htmx from 'htmx.org'.
> 
> Correct me if I'm wrong, but the current `export { htmx }` statement we have in the the `esm` wrapper supports this, right?

No, it does not. Again, this is fairy easy to reproduce:

Here is HTMX.org 2.0-alpha2
<img width="493" alt="Screenshot 2024-02-26 at 21 41 22" src="https://github.com/bigskysoftware/htmx/assets/1772890/1ac448d0-b3c1-4269-840a-004b95dd25f9">


Here is Lit, just to show how a module should look like in a browser.
<img width="556" alt="Screenshot 2024-02-26 at 21 41 45" src="https://github.com/bigskysoftware/htmx/assets/1772890/afda78dd-8eb9-407e-a40e-f1a25820e199">

**Comment by alexpetros** [comment]
Oh, I know exactly what you're talking about—you can't export anonymous _objects_ from an export statement i.e. the `export {}` just doesn't work. I completely forgot about that. Okay so the existing factory is bugged—we can fix that.

Is the change you were going to suggest just `export default htmx`?

**Comment by terrablue** [comment]
If this is implemented, it would be nice to add deep exports in `package.json` for the extensions such that it's possible to do something like `export * from "htmx.org/client-side-templates/handlebars";` to be picked up by the bundler, see https://github.com/terrablue/htmx-esm/blob/master/package.json#L11-L38

I would be more than happy to get rid of the needless vendoring, seeing as I need to update it every time a new HTMX version comes out.

**Comment by codingjoe** [comment]
> If this is implemented, it would be nice to add deep exports in `package.json` for the extensions such that it's possible to do something like `export * from "htmx.org/client-side-templates/handlebars";` to be picked up by the bundler, see https://github.com/terrablue/htmx-esm/blob/master/package.json#L11-L38
> 
> I would be more than happy to get rid of the needless vendoring, seeing as I need to update it every time a new HTMX version comes out.

Yes, I mentioned that in the original ticket report. I will submit a patch for that.

**Comment by codingjoe** [comment]
> Is the change you were going to suggest just `export default htmx`?

Yes, I believe this would be most consistent with the docs, where you call `htmx.someFunction`. Be it on a window or module scope. I will be easy copy and paste for most people.

**Comment by codingjoe** [comment]
Since there seems to be plenty of traffic on this issue, I will create a patch targeting the 2.0 release.

**Comment by codingjoe** [comment]
I looked at the code for 2.0 and decided to discontinue using HTMX. Maybe someone else will pick this up someday.

**Comment by alexpetros** [comment]
> I looked at the code for 2.0 and decided to discontinue using HTMX. Maybe someone else will pick this up someday.

Would you like to share what it was about the 2.0 code that discourage you? It's really not that different

**Comment by codingjoe** [comment]
> > I looked at the code for 2.0 and decided to discontinue using HTMX. Maybe someone else will pick this up someday.
> 
> Would you like to share what it was about the 2.0 code that discouraged you? It's really not that different

Hi @alexpetros, I would rather not discourage anyone. I know how much work goes into a project this size. Thus, will stick to errors:

1. extensions are not built in 2.0
2. test suite wasn't fully working (some files were renamed, but not their references)
3. extension are neither CSJ or ESM compatible

As a personal note: Tools like Wireit and Rollup or ESbuild might make the build processes more smooth. Lifting browser support to [Baseline 2023](https://web.dev/explore/baseline/2023) might also simplify things.

**Comment by alexpetros** [comment]
Ok! Well, anyway, we're going to switch to default exports for htmx 2.0: https://github.com/bigskysoftware/htmx/pull/2428
